### DISCUSSION TITLE
Kafka - Inspecting outgoing kafka messages.

### DISCUSSION THREAD

**Discussion by MintInfusion** [smallrye/smallrye-reactive-messaging#2300] [Q&A] [answered]
Hi, we are currently trying to automatically propagate an incoming HTTP Header in Quarkus as a kafka header in our microservice architecture for testing purposes.  Let's call that header "X-Testing-Data" for now.

My current plan is to use [channel decorators](https://smallrye.io/smallrye-reactive-messaging/4.9.0/concepts/decorators/) to do the following steps:

1. Inspect the outgoing kafka message, if it has already set "X-Testing-Data" as a kafka header, do nothing.
2. If the header is absent, see if "X-testing-data" is found in an HTTP message context and use that if available to populate the message.

Unfortunately, I am currently already failing with step 1. Outgoing messages seem to have varying formats, so to figure out where the Kafka Headers inside the message are,  I would have to reimplement a subset of this function here:

https://github.com/smallrye/smallrye-reactive-messaging/blob/8bfcb7f053615f5adc9825d2855bd5ab40978593/smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/impl/KafkaSink.java#L176-L249

But checking for the existence of certain classes inside of the message seems very likely to break in future updates. I did also not find anything about this in the API documentation.

Am I missing something? I assume adding/removing/manipulating key/value/header is the goal of channel decorators, but without a way to inspect the existing message, they seem unusable to me...

What I would have expected to see is a function that transforms any given message into a [ProducerRecord](https://kafka.apache.org/21/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html) or similar.

Hope someone can explain what I am failing to understand.

**Comment by ozangunalp** [comment, marked as answer]
If you want to propagate HTTP headers the best option is to use [OutgoingInterceptor](http://smallrye.io/smallrye-reactive-messaging/4.9.0/concepts/decorators/#intercepting-outgoing-messages) and add `OutgoingKafkaRecordMetadata` to your outgoing message. 

In Quarkus we recommend using [`ContextLocals`](https://quarkus.io/guides/duplicated-context#context-local-data) to propagate contextual information between HTTP and RM producer channels.

You can write a ServerRequestFilter to inject headers to be propagated to the `ContextLocals` and retrieve them in your OutgoingInterceptor. In normal use cases, the context is preserved.

I'll write a section with examples about this in the Quarkus Kafka guide.
